pkg/storage/stores/shipper/compactor/table_test.go

```diff
@@ -70,6 +71,103 @@ func TestTable_Compaction(t *testing.T) {
 }
 
+type TableMarkerFunc func(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error)
+
+func (t TableMarkerFunc) MarkForDelete(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error) {
+	return t(ctx, tableName, db)
+}
+
+func TestTable_CompactionRetention(t *testing.T) {
+	for name, tt := range map[string]struct {
+		dbCount     int
+		assert      func(t *testing.T, storagePath string)
+		tableMarker retention.TableMarker
+	}{
+		"emptied table": {
+			dbCount: 2,
+			assert: func(t *testing.T, storagePath string) {
+				_, err := os.Stat(storagePath)
+				require.True(t, os.IsNotExist(err))
+			},
+			tableMarker: TableMarkerFunc(func(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error) {
+				return true, 100, nil
+			}),
+		},
+		"marked table": {
+			dbCount: 2,
+			assert: func(t *testing.T, storagePath string) {
+				files, err := ioutil.ReadDir(storagePath)
+				require.NoError(t, err)
+				require.Len(t, files, 1)
+				require.True(t, strings.HasSuffix(files[0].Name(), ".gz"))
+			},
+			tableMarker: TableMarkerFunc(func(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error) {
+				return false, 100, nil
+			}),
+		},
+		"already compacted table": {
+			dbCount: 1,
+			assert: func(t *testing.T, storagePath string) {
+				files, err := ioutil.ReadDir(storagePath)
+				require.NoError(t, err)
+				require.Len(t, files, 1)
+				require.True(t, strings.HasSuffix(files[0].Name(), ".gz"))
+			},
+			tableMarker: TableMarkerFunc(func(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error) {
+				return false, 100, nil
+			}),
+		},
+		"not modified": {
+			dbCount: 1,
+			assert: func(t *testing.T, storagePath string) {
+				files, err := ioutil.ReadDir(storagePath)
+				require.NoError(t, err)
+				require.Len(t, files, 1)
+				require.True(t, strings.HasSuffix(files[0].Name(), ".gz"))
+			},
+			tableMarker: TableMarkerFunc(func(ctx context.Context, tableName string, db *bbolt.DB) (bool, int64, error) {
+				return false, 0, nil
+			}),
+		},
+	} {
+		tt := tt
+		t.Run(name, func(t *testing.T) {
+			tempDir := t.TempDir()
+
+			tableName := "test"
+			objectStoragePath := filepath.Join(tempDir, objectsStorageDirName)
+			tableWorkingDirectory := filepath.Join(tempDir, workingDirName, tableName)
+
+			// setup some dbs
+			numDBs := tt.dbCount
+			numRecordsPerDB := 100
+
+			dbsToSetup := make(map[string]testutil.DBRecords)
+			for i := 0; i < numDBs; i++ {
+				dbsToSetup[fmt.Sprint(i)] = testutil.DBRecords{
+					Start:      i * numRecordsPerDB,
+					NumRecords: (i + 1) * numRecordsPerDB,
+				}
+			}
+
+			testutil.SetupDBTablesAtPath(t, tableName, objectStoragePath, dbsToSetup, true)
+
+			// setup exact same copy of dbs for comparison.
+			testutil.SetupDBTablesAtPath(t, "test-copy", objectStoragePath, dbsToSetup, false)
```

Comment:
Looks like it is not used. Maybe you can use it in the `assert` to compare tables in `not modified` test.